aiida_testing/export_cache/_fixtures.py

```diff
@@ -0,0 +1,375 @@
+# -*- coding: utf-8 -*-
+"""
+Defines pytest fixtures for automatically enable caching in tests and export aiida graphs if not existent.
+Meant to be useful for WorkChain tests.
+"""
+# pylint: disable=unused-argument, protected-access, redefined-outer-name
+
+import os
+import hashlib
+import pathlib
+import typing as ty
+import pytest
+from aiida.engine import run_get_node  #,ProcessBuilder
+from aiida.common.hashing import make_hash
+from aiida.orm import Node, Code, load_node, Dict, SinglefileData, List, FolderData, RemoteData
+from aiida.orm import CalcJobNode, ProcessNode
+from aiida.orm.querybuilder import QueryBuilder
+from aiida.manage.caching import enable_caching
+from contextlib import contextmanager
+
+__all__ = (
+    "run_with_cache", "load_cache", "export_cache", "with_export_cache", "hash_code_by_entrypoint"
+)
+
+#### utils
+
+
+def unnest_dict(nested_dict: dict) -> dict:
+    """
+    Returns a simple dictionary from a possible arbitray nested dictionary
+    by adding keys in dot notation, rekrusively
+    """
+    new_dict = {}
+    for key, val in nested_dict.items():
+        if isinstance(val, dict):
+            unval = unnest_dict(val)  #rekursive!
+            for key2, val2 in unval.items():
+                key_new = str(key) + '.' + str(key2)
+                new_dict[key_new] = val2
+        else:
+            new_dict[str(key)] = val
+    return new_dict
+
+
+def get_hash_process(  # pylint: disable=dangerous-default-value
+    builder: dict,
+    input_nodes: list = []
+):
+    """ creates a hash from a builder/dictionary of inputs"""
+
+    # hashing the builder
+    # currently workchains are not hashed in AiiDA so we create a hash for the filename
+    unnest_builder = unnest_dict(builder)
+    md5sum = hashlib.md5()
+    for key, val in sorted(unnest_builder.items()):  # pylint: disable=unused-variable
+        if isinstance(val, Code):
+            continue  # we do not include the code in the hash, might be mocked
+            #TODO include the code to some extent
+        if isinstance(val, Node):
+            if not val.is_stored:
+                val.store()
+            val_hash = val.get_hash()  # only works if nodes are stored!
+            input_nodes.append(val)
+        else:
+            val_hash = make_hash(val)
+        md5sum.update(val_hash.encode())
+    bui_hash = md5sum.hexdigest()
+
+    return bui_hash, input_nodes
+
+
+####
+
+#### fixtures
+
+
+@pytest.fixture(scope='function')
+def export_cache(hash_code_by_entrypoint):
```

Comment:
I'm not sure if this should be a fixture (same for `load_cache`), or just an internal helper function. Do we want users to use the bare `export_cache`?

Reply:
I made this an extra fixture, because if we change the import/export functionality to be different from what aiida-core does, as a user I might want to use it in a test (or see what it does for debugging). In aiida-fleur tests I use a fixture to export and import things to create test environments, because we also have some input nodes for workchains where uuids are provided. So there might be a need to have something `bare` which is adapted for tests (like with specific hashing, or striped/slim exports etc.). For now this defines the interface. Overall I'm also not sure if this should be just a helper as for my current purposes do not need to take it further.